training/src/main/scala/com/airbnb/aerosolve/training/TrainingUtils.scala

```diff
@@ -213,80 +214,74 @@ object TrainingUtils {
     input
-      .mapPartitions(partition => {
-      // family, feature name => count, min, max, sum x, sum x ^ 2
-      val weights = new ConcurrentHashMap[(String, String), FeatureStatistics]().asScala
-      partition.foreach(examples => {
-        for (i <- 0 until examples.example.size()) {
-          val flatFeature = Util.flattenFeature(examples.example.get(i)).asScala
-          flatFeature.foreach(familyMap => {
-            familyMap._2.foreach(feature => {
-              val key = (familyMap._1, feature._1)
-              val curr = weights.getOrElse(key, FeatureStatistics(0, Double.MaxValue, Double.MinValue, 0.0, 0.0))
-                val v = feature._2
-                weights.put(key,
-                            FeatureStatistics(curr.count + 1,
-                             scala.math.min(curr.min, v),
-                             scala.math.max(curr.max, v),
-                             curr.mean + v, // actually the sum
-                             curr.variance + v * v) // actually the sum of squares
-                )
-              })
-          })
+      .flatMap(examples => {
+        examples.example.iterator().flatMap {
+          example =>
+            Util.flattenFeatureAsStream(example).iterator()
+              .flatMap {
+                featureFamily =>
+                  val family = featureFamily.getKey
+                  featureFamily.getValue.iterator().map {
+                    feature =>
+                      val value = feature.getValue.toDouble
+                      (family, feature.getKey, value)
+                  }
+              }
         }
       })
-      weights.iterator
-    })
-      .reduceByKey((a, b) =>
-                     FeatureStatistics(a.count + b.count,
-                      scala.math.min(a.min, b.min),
-                      scala.math.max(a.max, b.max),
-                      a.mean + b.mean,
-                      a.variance + b.variance))
-      .filter(x => x._2.count >= minCount)
-      .map(x => (x._1,
-          FeatureStatistics(
-           count = x._2.count,
-           min = x._2.min,
-           max = x._2.max,
-           mean = x._2.mean / x._2.count,
-           variance = (x._2.variance - x._2.mean * x._2.mean / x._2.count) / (x._2.count - 1.0)
-           )))
+      .toDF("family", "feature", "value")
+      .groupBy($"family", $"feature")
+      .agg(
+        count(lit(1)) as "count",
+        min($"value"),
+        max($"value"),
+        sum($"value"),
+        // TODO: use built-in stdev function in Spark 1.6
+        sum($"value" * $"value")
+      )
+      // ignore features present in less than minCount examples
+      .where($"count" >= minCount)
+      .map {
+        case Row(family: String, feature: String, count: Long, min: Double, max: Double, sum: Double, sqSum: Double) =>
+          ((family, feature), FeatureStatistics(count, min, max, sum / count, (sqSum - sum * sum / count) / (count - 1)))
```

Comment:
I think mean and/or variance will be not a number and such feature will probably be dropped anyway.